File: data/skills/wikijs/scripts/ppt_to_wiki.py
#!/usr/bin/env python3
"""
Convert PPT extracted content to Wiki.js pages.

Usage:
    python ppt_to_wiki.py \
        --content tmp/ppt_extracted/content.md \
        --images tmp/ppt_extracted/slides_png_hd2 \
        --output-path "ppt/presentation-name" \
        --title "Presentation Title"

Or with base64 embedded images:
    python ppt_to_wiki.py \
        --content tmp/ppt_extracted/content.md \
        --images tmp/ppt_extracted/slides_png_hd2 \
        --output-path "ppt/presentation-name" \
        --title "Presentation Title" \
        --embed-images
"""

import os
import sys
import base64
import argparse
from pathlib import Path

# Add scripts directory to path for importing wikijs_client
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from wikijs_client import WikiJSClient


def image_to_base64(image_path: str) -> str:
    """Convert image to base64 data URI."""
    with open(image_path, 'rb') as f:
        data = f.read()
        encoded = base64.b64encode(data).decode('utf-8')
    
    ext = Path(image_path).suffix.lower()
    mime_types = {
        '.png': 'image/png',
        '.jpg': 'image/jpeg',
        '.jpeg': 'image/jpeg',
        '.gif': 'image/gif',
        '.svg': 'image/svg+xml',
    }
    mime = mime_types.get(ext, 'image/png')
    return f'data:{mime};base64,{encoded}'


def create_images_section(images_dir: str, embed: bool = False) -> str:
    """Create markdown section with slide images."""
    if not images_dir or not os.path.exists(images_dir):
        return ""
    
    # Get all image files sorted
    image_files = sorted([
        f for f in os.listdir(images_dir)
        if f.lower().endswith(('.png', '.jpg', '.jpeg', '.gif', '.svg'))
    ])
    
    if not image_files:
        return ""
    
    section = f"## 📊 PPT幻灯片预览\n\n> 共 {len(image_files)} 页\n\n"
    
    for i, img_file in enumerate(image_files, 1):
        img_path = os.path.join(images_dir, img_file)
        
        if embed:
            # Embed as base64
            try:
                data_uri = image_to_base64(img_path)
                section += f"### 幻灯片 {i}\n\n![Slide {i}]({data_uri})\n\n"
            except Exception as e:
                section += f"### 幻灯片 {i}\n\n(Error loading image: {e})\n\n"
        else:
            # Just reference the image file
            section += f"### 幻灯片 {i}\n\n![Slide {i}]({img_file})\n\n"
        
        section += "---\n\n"
    
    return section


def convert_ppt_to_wiki(
    content_file: str,
    images_dir: str,
    output_path: str,
    title: str,
    embed_images: bool = False,
    locale: str = "en",
    tags: str = "ppt,presentation"
) -> dict:
    """
    Convert PPT extracted content to Wiki.js page.
    
    Args:
        content_file: Path to content.md file
        images_dir: Path to slides images directory
        output_path: Wiki.js page path (e.g., "ppt/my-presentation")
        title: Page title
        embed_images: Whether to embed images as base64
        locale: Page locale
        tags: Comma-separated tags
    
    Returns:
        API response dict
    """
    # Read content
    with open(content_file, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Build full page content
    images_section = create_images_section(images_dir, embed_images)
    
    full_content = f"""# {title}

> 本文档由PPT自动转换生成

{images_section}

## 📝 详细内容

{content}
"""
    
    # Create Wiki.js client
    client = WikiJSClient()
    
    # Create page
    tag_list = [t.strip() for t in tags.split(',') if t.strip()]
    
    result = client.create_page(
        path=output_path,
        title=title,
        content=full_content,
        locale=locale,
        description=f"PPT转换 - {title}",
        tags=tag_list,
        is_published=True,
        editor="markdown"
    )
    
    return result


def main():
    parser = argparse.ArgumentParser(
        description='Convert PPT extracted content to Wiki.js page'
    )
    parser.add_argument(
        '--content', '-c',
        required=True,
        help='Path to content.md file'
    )
    parser.add_argument(
        '--images', '-i',
        help='Path to slides images directory'
    )
    parser.add_argument(
        '--output-path', '-o',
        required=True,
        help='Wiki.js page path (e.g., "ppt/my-presentation")'
    )
    parser.add_argument(
        '--title', '-t',
        required=True,
        help='Page title'
    )
    parser.add_argument(
        '--embed-images', '-e',
        action='store_true',
        help='Embed images as base64 (creates self-contained page)'
    )
    parser.add_argument(
        '--locale', '-l',
        default='en',
        help='Page locale (default: en)'
    )
    parser.add_argument(
        '--tags',
        default='ppt,presentation',
        help='Comma-separated tags (default: ppt,presentation)'
    )
    
    args = parser.parse_args()
    
    # Validate inputs
    if not os.path.exists(args.content):
        print(f"Error: Content file not found: {args.content}")
        sys.exit(1)
    
    if args.images and not os.path.exists(args.images):
        print(f"Warning: Images directory not found: {args.images}")
        args.images = None
    
    print(f"Converting PPT to Wiki.js page...")
    print(f"  Content: {args.content}")
    print(f"  Images: {args.images or 'None'}")
    print(f"  Output: {args.output_path}")
    print(f"  Title: {args.title}")
    print(f"  Embed images: {args.embed_images}")
    
    result = convert_ppt_to_wiki(
        content_file=args.content,
        images_dir=args.images,
        output_path=args.output_path,
        title=args.title,
        embed_images=args.embed_images,
        locale=args.locale,
        tags=args.tags
    )
    
    # Check result
    response = result.get("data", {}).get("pages", {}).get("create", {})
    response_result = response.get("responseResult", {})
    
    if response_result.get("succeeded"):
        page = response.get("page", {})
        print(f"\n✅ Success!")
        print(f"   Page ID: {page.get('id')}")
        print(f"   Path: {page.get('path')}")
        print(f"   Title: {page.get('title')}")
    else:
        print(f"\n❌ Failed: {response_result.get('message')}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: data/skills/wikijs/scripts/ppt_to_wiki.js
#!/usr/bin/env node
/**
 * Convert PPT extracted content to Wiki.js pages
 * 
 * Usage:
 *   node ppt_to_wiki.js \
 *     --content tmp/ppt_extracted/content.md \
 *     --images tmp/ppt_extracted/slides_png_hd2 \
 *     --output-path "ppt/presentation-name" \
 *     --title "Presentation Title"
 * 
 * Or with asset upload:
 *   node ppt_to_wiki.js \
 *     --content tmp/ppt_extracted/content.md \
 *     --images tmp/ppt_extracted/slides_png_hd2 \
 *     --output-path "ppt/presentation-name" \
 *     --title "Presentation Title" \
 *     --upload-to-folder 1
 */

const fs = require('fs');
const path = require('path');
const WikiJSClient = require('./wikijs_client');

function parseArgs() {
    const args = process.argv.slice(2);
    const options = {
        content: null,
        images: null,
        outputPath: null,
        title: null,
        locale: 'en',
        tags: 'ppt,presentation',
        uploadToFolder: null,
        embedImages: false
    };

    for (let i = 0; i < args.length; i++) {
        switch (args[i]) {
            case '--content':
            case '-c':
                options.content = args[++i];
                break;
            case '--images':
            case '-i':
                options.images = args[++i];
                break;
            case '--output-path':
            case '-o':
                options.outputPath = args[++i];
                break;
            case '--title':
            case '-t':
                options.title = args[++i];
                break;
            case '--locale':
            case '-l':
                options.locale = args[++i];
                break;
            case '--tags':
                options.tags = args[++i];
                break;
            case '--upload-to-folder':
            case '-u':
                options.uploadToFolder = parseInt(args[++i]);
                break;
            case '--embed-images':
            case '-e':
                options.embedImages = true;
                break;
        }
    }

    return options;
}

function imageToBase64(imagePath) {
    const data = fs.readFileSync(imagePath);
    const ext = path.extname(imagePath).toLowerCase();
    const mimeTypes = {
        '.png': 'image/png',
        '.jpg': 'image/jpeg',
        '.jpeg': 'image/jpeg',
        '.gif': 'image/gif',
        '.svg': 'image/svg+xml'
    };
    const mime = mimeTypes[ext] || 'image/png';
    return `data:${mime};base64,${data.toString('base64')}`;
}

async function uploadImages(client, imagesDir, folderId) {
    const imageFiles = fs.readdirSync(imagesDir)
        .filter(f => /\.(png|jpg|jpeg|gif|svg)$/i.test(f))
        .sort();

    console.log(`Found ${imageFiles.length} images to upload...\n`);

    const results = [];
    for (let i = 0; i < imageFiles.length; i++) {
        const filename = imageFiles[i];
        const filePath = path.join(imagesDir, filename);
        
        process.stdout.write(`[${i + 1}/${imageFiles.length}] Uploading ${filename}... `);
        
        try {
            const result = await client.uploadFile(filePath, filename, folderId);
            if (result.succeeded || result === 'ok') {
                console.log('OK');
                results.push({ success: true, filename });
            } else {
                console.log(`Failed: ${result.message || result}`);
                results.push({ success: false, filename, error: result.message });
            }
        } catch (err) {
            console.log(`Error: ${err.message}`);
            results.push({ success: false, filename, error: err.message });
        }
        
        // Small delay to not overwhelm the server
        await new Promise(r => setTimeout(r, 300));
    }

    const successCount = results.filter(r => r.success).length;
    console.log(`\nUpload complete: ${successCount}/${results.length} successful`);
    
    return results;
}

function createImagesSection(imagesDir, options, uploadedImages = []) {
    if (!imagesDir || !fs.existsSync(imagesDir)) {
        return '';
    }

    const imageFiles = fs.readdirSync(imagesDir)
        .filter(f => /\.(png|jpg|jpeg|gif|svg)$/i.test(f))
        .sort();

    if (imageFiles.length === 0) {
        return '';
    }

    let section = `## 📊 PPT幻灯片预览\n\n> 共 ${imageFiles.length} 页\n\n`;

    for (let i = 0; i < imageFiles.length; i++) {
        const imgFile = imageFiles[i];
        section += `### 幻灯片 ${i + 1}\n\n`;

        if (options.embedImages) {
            // Embed as base64
            const imgPath = path.join(imagesDir, imgFile);
            try {
                const dataUri = imageToBase64(imgPath);
                section += `![Slide ${i + 1}](${dataUri})\n\n`;
            } catch (err) {
                section += `(Error loading image: ${err.message})\n\n`;
            }
        } else if (options.uploadToFolder) {
            // Reference uploaded asset
            section += `![Slide ${i + 1}](/ppt-images/${imgFile})\n\n`;
        } else {
            // Just reference the filename
            section += `![Slide ${i + 1}](${imgFile})\n\n`;
        }

        section += '---\n\n';
    }

    return section;
}

async function main() {
    const options = parseArgs();

    // Validate required options
    if (!options.content || !fs.existsSync(options.content)) {
        console.error('Error: Content file not found. Use --content to specify.');
        process.exit(1);
    }
    if (!options.outputPath) {
        console.error('Error: Output path required. Use --output-path to specify.');
        process.exit(1);
    }
    if (!options.title) {
        console.error('Error: Title required. Use --title to specify.');
        process.exit(1);
    }

    console.log('Converting PPT to Wiki.js page...');
    console.log(`  Content: ${options.content}`);
    console.log(`  Images: ${options.images || 'None'}`);
    console.log(`  Output: ${options.outputPath}`);
    console.log(`  Title: ${options.title}`);
    console.log(`  Upload to folder: ${options.uploadToFolder || 'None'}`);
    console.log(`  Embed images: ${options.embedImages}\n`);

    // Read content
    const content = fs.readFileSync(options.content, 'utf8');

    // Initialize client
    const client = new WikiJSClient();

    // Upload images if specified
    let uploadedImages = [];
    if (options.uploadToFolder && options.images) {
        uploadedImages = await uploadImages(client, options.images, options.uploadToFolder);
    }

    // Build images section
    const imagesSection = createImagesSection(options.images, options, uploadedImages);

    // Combine content
    const fullContent = `# ${options.title}\n\n> 本文档由PPT自动转换生成\n\n${imagesSection}\n## 📝 详细内容\n\n${content}`;

    console.log('\nCreating wiki page...');

    // Create page
    const result = await client.createPage({
        path: options.outputPath,
        title: options.title,
        content: fullContent,
        locale: options.locale,
        description: `PPT转换 - ${options.title}`,
        tags: options.tags.split(',').map(t => t.trim()),
        isPublished: true,
        editor: 'markdown'
    });

    const response = result.data?.pages?.create;
    
    if (response?.responseResult?.succeeded) {
        const page = response.page;
        console.log(`\n[SUCCESS] Page created!`);
        console.log(`  Page ID: ${page.id}`);
        console.log(`  Path: ${page.path}`);
        console.log(`  URL: https://wiki.erix.vip/en/${page.path}`);
    } else {
        console.log(`\n[FAILED] ${response?.responseResult?.message || 'Unknown error'}`);
        process.exit(1);
    }
}

main().catch(err => {
    console.error('Error:', err.message);
    process.exit(1);
});
